File: commands2/proxycommand.py
from .command import Command
from .commandgroup import *

from typing import Callable, overload, Union

class ProxyCommand(Command):

    _supplier: Callable[[], Command]

    @overload
    def __init__(self, supplier: Callable[[], Command]): ...

    @overload
    def __init__(self, command: Command): ...

    def __init__(self, *args, **kwargs):
        super().__init__()
        
        if "supplier" in kwargs:
            self._supplier = kwargs["supplier"]
        elif "command" in kwargs:
            self._supplier = lambda: kwargs["command"]
        elif len(args) == 1:
            if isinstance(args[0], Command):
                self._supplier = lambda: args[0]
            else:
                self._supplier = args[0]

    def initialize(self):
        self._command = self._supplier()
        self._command.schedule()

    def end(self, interrupted: bool):
        if interrupted and self._command is not None:
            self._command.cancel()
        self._command = None
    
    def execute(self):
        pass

    def isFinished(self) -> bool:
        return self._command is None or not self._command.isScheduled()
    
    def runsWhenDisabled(self) -> bool:
        return True

File: cpp/frc2/command/ProxyCommand.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc2/command/ProxyCommand.h"

#include <wpi/sendable/SendableBuilder.h>

using namespace frc2;

ProxyCommand::ProxyCommand(wpi::unique_function<Command*()> supplier)
    : m_supplier(std::move(supplier)) {}

ProxyCommand::ProxyCommand(wpi::unique_function<CommandPtr()> supplier)
    : ProxyCommand([supplier = std::move(supplier),
                    holder = std::optional<CommandPtr>{}]() mutable {
        holder = supplier();
        return holder->get();
      }) {}

ProxyCommand::ProxyCommand(Command* command)
    : ProxyCommand([command] { return command; }) {
  SetName(std::string{"Proxy("}.append(command->GetName()).append(")"));
}

ProxyCommand::ProxyCommand(std::unique_ptr<Command> command) {
  SetName(std::string{"Proxy("}.append(command->GetName()).append(")"));
  m_supplier = [command = std::move(command)] { return command.get(); };
}

void ProxyCommand::Initialize() {
  m_command = m_supplier();
  m_command->Schedule();
}

void ProxyCommand::End(bool interrupted) {
  if (interrupted) {
    m_command->Cancel();
  }
  m_command = nullptr;
}

bool ProxyCommand::IsFinished() {
  // because we're between `initialize` and `end`, `m_command` is necessarily
  // not null but if called otherwise and m_command is null, it's UB, so we can
  // do whatever we want -- like return true.
  return m_command == nullptr || !m_command->IsScheduled();
}

void ProxyCommand::InitSendable(wpi::SendableBuilder& builder) {
  Command::InitSendable(builder);
  builder.AddStringProperty(
      "proxied",
      [this] {
        if (m_command) {
          return m_command->GetName();
        } else {
          return std::string{"null"};
        }
      },
      nullptr);
}
